src/client/connection.rs

```
@@ -0,0 +1,97 @@
+use anyhow::Context;
+use anyhow::Result;
+use dap::base_message::Sendable;
+use dap::errors::ServerError;
+use dap::prelude::{Event, Request, ResponseBody, Server};
+use dap::server::{ServerReader, ServerWriter};
+use std::io::{BufReader, BufWriter};
+use std::net::{TcpListener, TcpStream};
+use std::sync::mpsc;
+use std::thread;
+
+pub struct Connection {
+    inbound_rx: mpsc::Receiver<Request>,
+    outbound_tx: mpsc::Sender<Sendable>,
+}
+
+impl Connection {
+    pub fn new() -> Result<Self> {
+        let tcp_listener = TcpListener::bind("127.0.0.1:0").map_err(ServerError::IoError)?;
+        let os_assigned_port = tcp_listener.local_addr()?.port();
+        // Print it so that the client can read it.
+        println!("\nDEBUGGER PORT: {os_assigned_port}");
+
+        let (stream, _client_addr) = tcp_listener.accept().map_err(ServerError::IoError)?;
+        let input = BufReader::new(stream.try_clone()?);
+        let output = BufWriter::new(stream);
+
+        let (server_reader, server_writer) = Server::new(input, output).split_server();
+
+        let (inbound_tx, inbound_rx) = mpsc::channel::<Request>();
+        let (outbound_tx, outbound_rx) = mpsc::channel::<Sendable>();
+
+        run_reader_thread(server_reader, inbound_tx);
+        run_writer_thread(server_writer, outbound_rx);
+
+        Ok(Self { inbound_rx, outbound_tx })
+    }
+
+    pub fn next_request(&self) -> Option<Request> {
+        self.inbound_rx.recv().ok()
+    }
+
+    pub fn try_next_request(&self) -> Option<Request> {
+        self.inbound_rx.try_recv().ok()
+    }
```

If we convert to option we won't know if it got disconnected (in which case we should bubble the error up/panic)